build_runner/bin/create_merged_dir.dart

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+
+import 'package:build_runner/src/asset/build_cache.dart';
+import 'package:build_runner/src/asset/file_based.dart';
+import 'package:build_runner/src/asset_graph/graph.dart';
+import 'package:build_runner/src/asset_graph/node.dart';
+import 'package:build_runner/src/package_graph/package_graph.dart';
+import 'package:build_runner/src/util/constants.dart';
+
+AssetGraph assetGraph;
+PackageGraph packageGraph;
+
+Future main(List<String> args) async {
+  stdout.writeln(
+      'Warning: this tool is unsupported and usage may change at any time, '
+      'use at your own risk.\n\n'
+      'This tool also assumes you are using the `writeToCache=true` option.');
+
+  var parsedArgs = argParser.parse(args);
+
+  var scriptPath = parsedArgs['script'] as String;
+  if (scriptPath == null) {
+    throw new ArgumentError.notNull('--script');
+  }
+
+  var outputDir = new Directory(parsedArgs['output-dir'] as String);
+  if (outputDir.existsSync()) {
+    stdout.writeln('Deleting existing output dir `$outputDir`');
+    outputDir.deleteSync(recursive: true);
+  }
+  outputDir.createSync(recursive: true);
+
+  var scriptFile = new File(scriptPath);
+  if (!scriptFile.existsSync()) {
+    throw new ArgumentError(
+        'Expected a build script at $scriptPath but didn\'t find one.');
+  }
+
+  var assetGraphFile = new File(assetGraphPathFor(p.absolute(scriptPath)));
+  if (!assetGraphFile.existsSync()) {
+    throw new ArgumentError(
+        'Unable to find AssetGraph for $scriptPath at ${assetGraphFile.path}');
+  }
+  stdout.writeln('Loading asset graph at ${assetGraphFile.path}...');
+
+  assetGraph = new AssetGraph.deserialize(
+      JSON.decode(assetGraphFile.readAsStringSync()) as Map);
+
+  packageGraph = new PackageGraph.forThisPackage();
+
+  stdout.writeln('Creating output dir at ${outputDir.path}...');
+  var rootDirs = new Set<String>();
+  for (var node in assetGraph.packageNodes(packageGraph.root.name)) {
+    var parts = p.url.split(node.id.path);
+    if (parts.length == 1) continue;
+    var dir = parts.first;
+    if (dir == outputDir.path) continue;
+    rootDirs.add(parts.first);
+  }
+
+  var reader = new BuildCacheReader(new FileBasedAssetReader(packageGraph),
+      assetGraph, packageGraph.root.name);
+
+  for (var node in assetGraph.allNodes) {
+    if (node is SyntheticAssetNode) continue;
+    if (node is GeneratedAssetNode && !node.wasOutput) continue;
+    if (node.id.path == '.packages') continue;
```

Comment:
should we be excluding this from ever getting to the asset graph?

Reply:
Yes actually, because its never really valid once we do a build.

Filed https://github.com/dart-lang/build/issues/668